oi/ContainerInfo.cpp

```diff
@@ -164,10 +165,84 @@ std::unique_ptr<ContainerInfo> ContainerInfo::loadFromFile(
       std::move(replaceTemplateParamIndex),
       allocatorIndex,
       underlyingContainerIndex,
-
+      {},
       {
           std::move(decl),
           std::move(func),
       },
   });
 }
+
+ContainerInfo::ContainerInfo(const fs::path& path) {
```

Comment:
I'm personally not a fan of this constructor, but that could be a holdover from other languages. Particularly the:
`const auto &info = containerInfos_.emplace_back(path);` which comes later is very unnatural to me.

Does this approach have significant benefits over changing the signature of the previous to `ContainerInfo ContainerInfo::loadFromFile(...)`? That would also give us the opportunity to use the more explicit `std::expected<ContainerInfo>` in C++23.

It's not a hill I want to die on, but a non-explicit and exception throwing cast from `fs::path&` to `ContainerInfo` feels strange to me.

Reply:
There are two separate things at play here:
1. Initialising ContainerInfo in the constructor

    This is idiomatic C++. After calling the ctor it is expected that an object will be fully usable. I would say the the invariant of a ContainerInfo class should be that it represents a container, i.e. has a valid typeName, ctype, etc. There are a few core guidelines I found on this: [C.40](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-ctor) [C.41](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-complete) [C.42](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-throw) [C.50](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-factory).

    This method is also more flexible and more efficient, as it means that ContainerInfo objects can be constructed anywhere, rather than forcing users into having them inside unique_ptrs.

1. Using std::vector's `emplace_back` method

    The new constructor is marked "explicit", so it won't allow implicit casts from a file path to ContainerInfo. The apparent implicitness comes from using `emplace_back`, which forwards the argument from this function call onto ContainerInfo's ctor.

    `v.emplace_back(path)` is similar to writing `v.push_back(ContainerInfo(path))`, except it guarantees that the object will be constructed in place, so is more efficient.